reactioncommerce/reaction
# Framework for Shopify synchronization + Basic Inventory sync

**This PR aims to do the following**  
* Provide an example for how to use webhooks in a plugin
* Extend the Shopify connector to be capable of syncing data from a Shopify store to a Reaction store.
* Provide a framework that sets up webhooks when a sync is started and removes the webhooks when the sync is turned off
* Provides a basic example sync which syncs inventory from Shopify to Reaction when an order is placed on Shopify

**This PR contains a new idea that should be reviewed extra carefully**
* Event logging on products and variants. Used to note that an external service reduced the inventory of an item, but is an informal proposal for how to log more data across Reaction.
  * New Schema file [schema/logs.js](https://github.com/reactioncommerce/reaction/blob/spencer-sync-shopify-orders/lib/collections/schemas/logs.js)
  * Used to note that inventory was updated for a particular variant in [sync/orders.js](https://github.com/reactioncommerce/reaction/blob/spencer-sync-shopify-orders/imports/plugins/included/connectors-shopify/server/methods/sync/orders.js#L62-L68)

**To Test:**  
Unfortunately, there's not an easy way to test this PR without deploying it to a real domain, or modifying some code and using a tunnel such as [ngrok](https://ngrok.com/).

1. Pull this branch & reset
2. Add Shopify Credentials.
3. Select Products and click Import - wait until the import is finished (You shouldn't need to wait, but it makes testing consistent)
4. Fire Up [ngrok](https://ngrok.com/) and determine what domain is tunneling to your localhost
you should have a terminal that looks something like this:
![image](https://user-images.githubusercontent.com/1203639/30260490-d9ec45f4-9684-11e7-85e5-d75342fdf9ac.png)
you'll see two urls `http://_______.ngrok.io` and `https://_______.ngrok.io` you want to use the `http` version as Shopify checks certificates on https webhooks.

5. **EDIT SOME CODE** - on line [48/49 of sync.js](https://github.com/reactioncommerce/reaction/blob/spencer-sync-shopify-orders/imports/plugins/included/connectors-shopify/server/methods/sync/sync.js#L48) - after integrations, update the Meteor call to look like this
*Note the slash `/` at the end of the ngrok url, it's necessary.*
    ```
      Meteor.call("connectors/shopify/createWebhook", {
        topic: topic,
        integrations: integrationsByTopic[topic],
        absoluteUrl: "http://yourdomain.ngrok.io/"
      });
    ```
    *Without adding this `absoluteUrl`, Shopify will attempt to add a webhook for your actual absoluteUrl which is `localhost` which will cause a failure as Shopify's webhook API will reject any attempt to create a webhook for a `localhost` url*
6. Select `Update Inventory when new Shopify order is placed` and click Sync to create webhooks webhooks.
7. Select a product that you want to test that's on both the Shopify store and the Reaction store. Note the inventory quantity for a particular variant or option.
8. Checkout with that particular variant or option.
9. The inventory available on the Reaction site should be reduced by the number of inventory that you checked out with on Shopify.
